### PR TITLE
JClouds Profitbricks provider - LoadBalancer API

### PR DESCRIPTION
**Pull request by jasminSPC**
JClouds Profitbricks provider - LoadBalancer API

### CODE REVIEW

**Comment by nacx** [comment]
This PR has several commits, including a merge commit that will make it difficult to merge. Before we start reviewing (and adding commits that address the review comments), could you cleanup the PR so it has only the necessary commits, and does not have any merge commit? This will facilitate the subsequent review commits/squashes, etc.

**Comment by jasminSPC** [comment]
Ping

File: profitbricks/src/main/java/org/jclouds/profitbricks/binder/loadbalancer/DeregisterLoadBalancerRequestBinder.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      requestBuilder.append("<ws:deregisterServersOnLoadBalancer>")
+              .append("<request>");
+      for (String s : payload.serverIds()) {
+         requestBuilder.append(format("<serverIds>%s</serverIds>", s));
+      }
+      requestBuilder.append(format("<loadBalancerId>%s</loadBalancerId>", payload.id()))
+              .append("</request>")
+              .append("</ws:deregisterServersOnLoadBalancer>");
+
+      return requestBuilder.toString().replaceAll("\\s+", "");
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/binder/loadbalancer/DeregisterLoadBalancerRequestBinder.java]: Is this really needed?

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+      ROUND_ROBIN, UNRECOGNIZED;
+
+      public static LoadbalancerAlgorithm
+              fromValue(String value) {
+         try {
+            return valueOf(value);
+         } catch (IllegalArgumentException e) {
+            return UNRECOGNIZED;
+         }
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: [nit] There is also a cleaner approach using Guava:
```java
return Enums.getIfPresent(LoadbalancerAlgorithm.class, value).or(UNRECOGNIZED);
```

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+   public abstract LoadbalancerAlgorithm loadBalancerAlgorithm();
+
+   @Nullable
+   public abstract String dataCenterId();
+
+   @Nullable
+   public abstract String dataCenterVersion();
+
+   @Nullable
+   public abstract boolean internetAccess();
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Primitive types can't be nullable

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+   public abstract Date creationTime();
+
+   @Nullable
+   public abstract Date lastModificationTime();
+
+   @Nullable
+   public abstract List<Server> balancedServers();
+
+   @Nullable
+   public abstract List<Firewall> firewalls();
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Are **all** fields of the LoadBalancer really nullable? Is it possible that there is a load balancer without name, without id, algorithm, etc? If not, configure `@Nullable` annoations properly.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+      public boolean internetAccess;
+
+      public String ip;
+
+      public String lanId;
+
+      public ProvisioningState state;
+
+      @Nullable
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Don't use this in the builder. Builder properties are not meant to be read.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+         public abstract String loadBalancerName();
+
+         public abstract LoadbalancerAlgorithm loadBalancerAlgorithm();
+
+         public abstract String ip();
+
+         @Nullable
+         public abstract String lanId();
+
+         @Nullable
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Avoid null lists and make sure they're immutable.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+            public CreatePayload build() {
+               return CreatePayload.create(dataCenterId, loadBalancerName, loadBalancerAlgorithm, ip, lanId, serverIds);
+            }
+         }
+      }
+
+      @AutoValue
+      public abstract static class RegisterPayload {
+
+         public abstract List<String> serverIds();
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Same comment about immutable lists.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+            public RegisterPayload build() {
+               return RegisterPayload.create(serverIds, id);
+            }
+         }
+      }
+
+      @AutoValue
+      public abstract static class DeregisterPayload {
+
+         public abstract List<String> serverIds();
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Same here.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/Server.java [hunk 9]
```diff
@@ -84,14 +85,20 @@ public static Status fromValue(String v) {
    @Nullable
    public abstract List<Nic> nics();
 
+   @Nullable
+   public abstract String balancedNicId();
+
+   @Nullable
+   public abstract boolean activate();
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/Server.java]: Primitives can't be null.

File: profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/BaseLoadBalancerResponseHandler.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+   protected final ServerListResponseHandler balancedServerResponseHandler;
+   protected final FirewallListResponseHandler firewallListResponseHandler;
+
+   protected LoadBalancer.Builder builder;
+   protected final DateCodec dateCodec;
+
+   protected boolean useBalancedServerParser = false;
+   protected boolean useFirewallParser = false;
+
+   protected BaseLoadBalancerResponseHandler(DateCodecFactory dateCodec, ServerListResponseHandler balancedServerResponseHandler, FirewallListResponseHandler firewallResponseHandler) {
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/BaseLoadBalancerResponseHandler.java]: Add null checks for the mandatory parameters. Although subclasses use injection, this class does not know how subclasses are built, so we'd better check the input parameters here.

File: profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/LoadBalancerListResponseHandler.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import org.jclouds.profitbricks.http.parser.firewall.FirewallListResponseHandler;
+import org.jclouds.profitbricks.http.parser.server.ServerListResponseHandler;
+import org.xml.sax.SAXException;
+
+public class LoadBalancerListResponseHandler extends BaseLoadBalancerResponseHandler<List<LoadBalancer>> {
+
+   private final List<LoadBalancer> loadBalancers;
+
+   @Inject
+   protected LoadBalancerListResponseHandler(DateCodecFactory dateCodec, ServerListResponseHandler balancedServerResponseHandler, FirewallListResponseHandler firewallListResponseHandler) {
```
Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/LoadBalancerListResponseHandler.java]: Make this package private.

File: profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+import static org.testng.Assert.assertNotNull;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "LoadbalancerApiLiveTest")
+public class LoadbalancerApiLiveTest extends BaseProfitBricksLiveTest {
+
+   String dataCenterId;
+   String loadBalancerID;
+   String serverId;
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java]: Any reason why these are not private?

File: profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+      System.out.println(loadBalancer);
+      assertNotNull(loadBalancer);
+
+   }
+
+   @Test
+   public void testGetAllLoadBalancers() {
+      List<LoadBalancer> loadBalancers = api.loadBalancerApi().getAllLoadBalancers();
+
+      assertNotNull(loadBalancers);
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java]: The API should never return null lists. Remove this.
Also, should this be executed after creating the load balancer?

File: profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+      assertFalse(loadBalancers.isEmpty());
+   }
+
+   @Test(dependsOnMethods = "testCreateLoadBalancer")
+   public void testGetLoadBalancer() {
+      LoadBalancer loadBalancer = api.loadBalancerApi().getLoadBalancer(loadBalancerID);
+
+      System.out.println(loadBalancer);
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java]: Remove this

File: profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+              .loadBalancerName("testName")
+              .loadBalancerAlgorithm(LoadbalancerAlgorithm.ROUND_ROBIN)
+              .ip("0.0.0.1")
+              .lanId("1")
+              .serverIds(serverIds)
+              .build();
+
+      LoadBalancer loadBalancer = api.loadBalancerApi().createLoadBalancer(payload);
+
+      System.out.println(loadBalancer);
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiLiveTest.java]: Remove this

File: profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiMockTest.java [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+         LoadBalancer loadBalancer = api.getLoadBalancer(id);
+
+         assertRequestHasCommonProperties(server.takeRequest(), content);
+         assertNotNull(loadBalancer);
+         assertEquals(loadBalancer.id(), id);
+      } finally {
+         pbApi.close();
+         server.shutdown();
+      }
+   }
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiMockTest.java]: Both get methods define a fallback. Add the corresponding tests.

File: profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiMockTest.java [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+      MockWebServer server = mockWebServer();
+      server.enqueue(new MockResponse().setBody(payloadFromResource("/loadbalancer/loadbalancers.xml")));
+
+      ProfitBricksApi pbApi = api(server.getUrl(rootUrl));
+      LoadBalancerApi api = pbApi.loadBalancerApi();
+
+      try {
+         List<LoadBalancer> loadBalancerList = api.getAllLoadBalancers();
+
+         assertRequestHasCommonProperties(server.takeRequest());
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/features/LoadbalancerApiMockTest.java]: Verify the body of the request too.

File: profitbricks/src/test/java/org/jclouds/profitbricks/http/parser/loadbalancer/LoadBalancerListResponseHandlerTest.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+import org.jclouds.profitbricks.domain.LoadBalancer;
+import org.jclouds.profitbricks.domain.LoadBalancer.LoadbalancerAlgorithm;
+import org.jclouds.profitbricks.domain.ProvisioningState;
+import org.jclouds.profitbricks.domain.Server;
+import org.jclouds.profitbricks.http.parser.BaseResponseHandlerTest;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "LoadbalancerListResponseHandlerTest")
```
Comment (nacx) [inline profitbricks/src/test/java/org/jclouds/profitbricks/http/parser/loadbalancer/LoadBalancerListResponseHandlerTest.java]: [nit] "LoadBalancer"

**Comment by nacx** [comment]
Add the missing tests for the `DeregisterLoadBalancerRequestBinder` and `UpdateLoadBalancerRequestBinder`, and update the Firewall and Server list handler tests to properly test the new elements introduced in this commit.

File: profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java (hunk 4 above)
Comment (jasminSPC) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: I use the same object for as a response when creating https://github.com/StackPointCloud/jclouds-labs/commit/42a6251088b7aa403407676e90ee70cf65a97503#diff-bf57b5d122ac037379a25c1825cf7badR65

Do you think I should change this to return RequestId only?

Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/domain/LoadBalancer.java]: Yes, I would change that. Think about how a user will interact with that API. What can be done with the returned object after it is created? Nothing. Then we'd better return the request id and let the user get the right object.

File: profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/BaseLoadBalancerResponseHandler.java (hunk 10 above)
Comment (jasminSPC) [inline profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/BaseLoadBalancerResponseHandler.java]: Should I throw exception if null? I don't see that we are doing the same here https://github.com/StackPointCloud/jclouds-labs/blob/master/profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/server/BaseServerResponseHandler.java#L50

Comment (nacx) [inline profitbricks/src/main/java/org/jclouds/profitbricks/http/parser/loadbalancer/BaseLoadBalancerResponseHandler.java]: Yes, I'd do it in both places, and I'd remove the `@Inject` annotation from the one you reference.

**Comment by jasminSPC** [comment]
ping

**Comment by baldwinSPC** [comment]
@nacx Just following up on this PR. Is there any outstanding work on our end we should do?

**Comment by nacx** [comment]
Apologies for the absence of feedback. We are about to release jclouds 1.9.0, and the master branch are under code freeze. We are fixing an issue spotted while testing the first release candidate, and expect to have the release out by the end of this week. Once that happens I'll be back to the PR backlog. Apologies for the delay!

**Comment by baldwinSPC** [comment]
@nacx No worries. Thanks. Good luck with the 1.9.0 release!

**Comment by nacx** [comment]
Hi! 1.9.0 is out and we're back to reviewing PRs. I've seen there are several files with Windows line endings. Mind fixing that?

**Comment by jasminSPC** [comment]
@nacx Last week I fixed those line endings. Is there something I missed or should I squash it?

**Comment by nacx** [comment]
> and update the Firewall and Server list handler tests to properly test the new elements introduced in this commit

The server object now seems to have nested load balancers, and the handler was updated to handle that. Update the handler tests to exercise that path too.

**Comment by nacx** [comment]
Just that minor and it will be ready to be merged. Apologies for the delay in the reviews!

**Comment by jasminSPC** [comment]
@nacx I don't see any changes on Firewalls that I have made. I only used Firewalls as nested objects in LoadBalancers. Am I missing something?

**Comment by nacx** [comment]
Looked at it again and it was only the Server object.
Pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/823df74f) and [1.9.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/c607a19a). Thanks!

**Pull request closed without merging**